README.md
StatsD InfluxDB backend
-----------------------

A naive [InfluxDB](http://influxdb.org/) backend for
[StatsD](https://github.com/etsy/statsd).

It can ship events to InfluxDB using two different strategies which can be
used at the same time.

### Regular Flush Strategy

StatsD will flush aggregated metrics with a configured interval. This is
the regular StatsD mode of operation.

### Proxy Strategy

This will map every incoming StatsD packet to an InfluxDB event. It's useful
if you want to store the raw events in InfluxDB without any rollups.

## CAVEATS

This is pretty young and I do not have much experience with InfluxDB yet.
Especially the event buffering and the event mapping might be problematic
and inefficient.

InfluxDB is also pretty young and there might be breaking changes until it
reaches 1.0.

Please be careful!

## Installation

    $ cd /path/to/statsd
    $ npm install statsd-influxdb-backend

## Configuration

You can configure the following settings in your StatsD config file.

```js
{
  graphitePort: 2003,
  graphiteHost: "graphite.example.com",
  port: 8125,
  backends: [ "./backends/graphite", "statsd-influxdb-backend" ],

  influxdb: {
    host: '127.0.0.1',   // InfluxDB host. (default 127.0.0.1)
    port: 8086,          // InfluxDB port. (default 8086)
    version: 0.8,        // InfluxDB version. (default 0.8) (options 0.8,0.9)
    ssl: false,          // InfluxDB is hosted over SSL. (default false)
    database: 'dbname',  // InfluxDB database instance. (required)
    username: 'user',    // InfluxDB database username.
    password: 'pass',    // InfluxDB database password.
    flush: {
      enable: true       // Enable regular flush strategy. (default true)
    },
    proxy: {
      enable: false,       // Enable the proxy strategy. (default false)
      suffix: 'raw',       // Metric name suffix. (default 'raw')
      flushInterval: 1000  // Flush interval for the internal buffer.
                           // (default 1000)
    },
    includeStatsdMetrics: false, // Send internal statsd metrics to InfluxDB. (default false)
    includeInfluxdbMetrics: false // Send internal backend metrics to InfluxDB. (default false)
                                  // Requires includeStatsdMetrics to be enabled.
  }
}
```

## Activation

Add the `statsd-influxdb-backend` to the list of StatsD backends in the config
file and restart the StatsD process.

```js
{
  backends: ['./backends/graphite', 'statsd-influxdb-backend']
}
```

## Unsupported Metric Types

#### Proxy Strategy

* Counter with sampling.
* Signed gauges. (i.e. `bytes:+4|g`)
* Sets

## InfluxDB Event Mapping

StatsD packets are currently mapped to the following InfluxDB events. This is
a first try and I'm open to suggestions to improve this.

### Set

StatsD package `client_version:1.1|c`, `client_version:1.2|c` as Influx event:

```js
[
  {
    name: 'visior',
    columns: ['value', 'time'],
    points:  [['1.1', 1384798553000], ['1.2', 1384798553001]]
  }
]
```

If you are using Grafana to visualize a Set, then using this query or
something similar

```
SELECT version, count(version) FROM client_version GROUP BY version, time(1m)
```

Also, to count for the size of unique value, another InfluxDB event is
also pushed

```js
[
  {
    name: 'visitor_count',
    columns: ['value', 'time'],
    points:  [set.length, 1384798553001]
  }
]
```

### Counter

StatsD packet `requests:1|c` as InfluxDB event:

#### Flush Strategy

```js
[
  {
    name: 'requests.counter',
    columns: ['value', 'time'],
    points: [[802, 1384798553000]]
  }
]
```

#### Proxy Strategy

```js
[
  {
    name: 'requests.counter.raw',
    columns: ['value', 'time'],
    points: [[1, 1384472029572]]
  }
]
```

### Timing

StatsD packet `response_time:170|ms` as InfluxDB event:

#### Flush Strategy

```js
[
  {
    name: 'response_time.timer.mean_90',
    columns: ['value', 'time'],
    points: [[445.25761772853184, 1384798553000]]
  },
  {
    name: 'response_time.timer.upper_90',
    columns: ['value', 'time'],
    points: [[905, 1384798553000]]
  },
  {
    name: 'response_time.timer.sum_90',
    columns: ['value', 'time'],
    points: [[321476, 1384798553000]]
  },
  {
    name: 'response_time.timer.std',
    columns: ['value', 'time'],
    points: [[294.4171159604542, 1384798553000]]
  },
  {
    name: 'response_time.timer.upper',
    columns: ['value', 'time'],
    points: [[998, 1384798553000]]
  },
  {
    name: 'response_time.timer.lower',
    columns: ['value', 'time'],
    points: [[2, 1384798553000]]
  },
  {
    name: 'response_time.timer.count',
    columns: ['value', 'time'],
    points: [[802, 1384798553000]]
  },
  {
    name: 'response_time.timer.count_ps',
    columns: ['value', 'time'],
    points: [[80.2, 1384798553000]]
  },
  {
    name: 'response_time.timer.sum',
    columns: ['value', 'time'],
    points: [[397501, 1384798553000]]
  },
  {
    name: 'response_time.timer.mean',
    columns: ['value', 'time'],
    points: [[495.6371571072319, 1384798553000]]
  },
  {
    name: 'response_time.timer.median',
    columns: ['value', 'time'],
    points: [[483, 1384798553000]]
  }
]
```

#### Proxy Strategy

```js
[
  {
    name: 'response_time.timer.raw',
    columns: ['value', 'time'],
    points: [[170, 1384472029572]]
  }
]
```

### Gauges

StatsD packet `bytes:123|g` as InfluxDB event:

#### Flush Strategy

```js
[
  {
    name: 'bytes.gauge',
    columns: ['value', 'time'],
    points: [[123, 1384798553000]]
  }
]
```

#### Proxy Strategy

```js
[
  {
    name: 'bytes.gauge.raw',
    columns: ['value', 'time'],
    points: [['gauge', 123, 1384472029572]]
  }
]
```

## Proxy Strategy Notes

### Event Buffering

To avoid one HTTP request per StatsD packet, the InfluxDB backend buffers the
incoming events and flushes the buffer on a regular basis. The current default
is 1000ms. Use the `influxdb.proxy.flushInterval` to change the interval.

This might become a problem with lots of incoming events.

The payload of a HTTP request might look like this:

```js
[
  {
    name: 'requests.counter.raw',
    columns: ['value', 'time'],
    points: [
      [1, 1384472029572],
      [1, 1384472029573],
      [1, 1384472029580]
    ]
  },
  {
    name: 'response_time.timer.raw',
    columns: ['value', 'time'],
    points: [
      [170, 1384472029570],
      [189, 1384472029572],
      [234, 1384472029578],
      [135, 1384472029585]
    ]
  },
  {
    name: 'bytes.gauge.raw',
    columns: ['value', 'time'],
    points: [
      [123, 1384472029572],
      [123, 1384472029580]
    ]
  }
]
```

## Backend Metrics

The following internal metrics are calculated for each flush:

- `statsd.influxdbStats.flush_time` - Time taken to process a complete flush in ms. Excluding the asynchronous HTTP Post.
- `statsd.influxdbStats.http_response_time` - Response time in ms of the InfluxDB HTTP endpoint when POSTing data.
- `statsd.influxdbStats.payload_size` - The size in bytes of the JSON payload.
- `statsd.influxdbStats.num_stats` - The number of metrics sent to InfluxDB in the last flush.

These are added to the set of internal statsd metrics. If both `influxdb.includeStatsdMetrics` and `influxdb.includeInfluxdbMetrics` are enabled, then these will be sent to InfluxDB when using the flush strategy.

The internal metrics can also can be viewed using the `stats` command on the [StatsD TCP Admin Interface](https://github.com/etsy/statsd/blob/master/docs/admin_interface.md)

## Contributing

All contributions are welcome: ideas, patches, documentation, bug reports,
complaints, and even something you drew up on a napkin.


CHANGELOG.md
StatsD InfluxDB backend - CHANGELOG
-----------------------------------

## v0.6.0 (2015-06-23)

* Initial InfluxDB 0.9 API support (#14, #17)
* Make username and password config options optional. (#15)

## v0.5.0 (2015-04-02)

* Unbreak conditional. (#13)
* Updated regex for statsd metrics. (#12)
* Expose internal backend metrics. (#9)

## v0.4.1 (2015-02-23)

* Use prefixStats setting instead of hardcoding prefix. (#6)

## v0.4.0 (2015-02-22)

* Add configuration option to enable sending internal statsd metrics. (#6)
* Add SET support. (#7)
* Fix problem with histograms breaking data flushes. (#8)
* Add SSL support. (#3)
* Improve configuration example in README.

## v0.3.0 (2014-08-24)

* Allow gauges with a value of 0 to be sent. (#2)
* Update `time_precision` to use `ms`. (#1)

## v0.2.0 (2013-11-19)

* Add flush strategy and enable it by default.
* Add configuration options for flush and proxy strategy.

## v0.1.0 (2013-11-15)

* Initial release.


scripts/setup-influxdb.js
var http = require('http'),
    querystring = require('querystring');

function httpRequest(method, path, payload) {
  var query = {u: 'root', p: 'root'};

  var req = http.request({
    hostname: 'localhost',
    port: 8086,
    path: path + '?' + querystring.stringify(query),
    method: method.toUpperCase(),
    agent: false
  });

  req.on('response', function (res) {
    res.on('data', function (chunk) {
      console.log('RESPONSE: ' + chunk.toString());
    });
  });

  req.on('error', function (e) {
    console.log(e);
  });

  if (payload) {
    req.write(JSON.stringify(payload));
  }
  req.end();
}

switch (process.argv[2]) {
  case 'createdb': {
    console.log('Creating database: "statsd"');
    httpRequest('post', '/db', {name: "statsd"});
    break;
  }
  case 'dropdb': {
    console.log('Deleting database: "statsd"');
    httpRequest('delete', '/db/statsd');
    break;
  }
  case 'createuser': {
    console.log('Creating user "user" with password "pass"');
    httpRequest('post', '/db/statsd/users', {username: 'user', password: 'pass'});
    break;
  }
  case 'createadmin': {
    console.log('Creating user "admin" with password "pass"');
    httpRequest('post', '/db/statsd/users', {username: 'admin', password: 'pass', admin: true});
    break;
  }
  default: {
    console.log('Commands: createdb, dropdb, createuser');
  }
}


lib/influxdb.js
/*
 * Flush stats to InfluxDB (http://influxdb.org/)
 *
 * To enable this backend, include 'statsd-influxdb-backend' in the backends
 * configuration array:
 *
 *   backends: ['statsd-influxdb-backend']
 *
 * The backend will read the configuration options from the following
 * 'influxdb' hash defined in the main statsd config file:
 *
 * influxdb: {
 *   host: '127.0.0.1',   // InfluxDB host. (default 127.0.0.1)
 *   port: 8086,          // InfluxDB port. (default 8086)
 *   ssl: false,          // InfluxDB is hosted over SSL. (default false)
 *   database: 'dbname',  // InfluxDB database instance. (required)
 *   version: 0.8,        // InfluxDB version. (default 0.8) (options 0.8,0.9)
 *   username: 'user',    // InfluxDB database username.
 *   password: 'pass',    // InfluxDB database password.
 *   flush: {
 *     enable: true       // Enable regular flush strategy. (default true)
 *   },
 *   proxy: {
 *     enable: false,       // Enable the proxy strategy. (default false)
 *     suffix: 'raw',       // Metric name suffix. (default 'raw')
 *     flushInterval: 1000  // Flush interval for the internal buffer.
 *                          // (default 1000)
 *   },
 *   includeStatsdMetrics: false, // Send internal statsd metrics to InfluxDB. (default false)
 *   includeInfluxdbMetrics: false // Send internal backend metrics to InfluxDB. (default false)
 *                                 // Requires includeStatsdMetrics to be enabled.
 * }
 *
 */
var util = require('util'),
	querystring = require('querystring'),
	http = require('http'),
	https = require('https');

function InfluxdbBackend(startupTime, config, events) {
	var self = this;

	self.debug = config.debug;
	self.registry = {};
	self.influxdbStats = {};

	self.defaultHost = '127.0.0.1';
	self.defaultPort = 8086;
	self.defaultVersion = 0.8;
	self.defaultFlushEnable = true;
	self.defaultProxyEnable = false;
	self.defaultProxySuffix = 'raw';
	self.defaultProxyFlushInterval = 1000;

	self.host = self.defaultHost;
	self.port = self.defaultPort;
	self.version = self.defaultVersion;
	self.protocol = http;
	self.flushEnable = self.defaultFlushEnable;
	self.proxyEnable = self.defaultProxyEnable;
	self.proxySuffix = self.defaultProxySuffix;
	self.proxyFlushInterval = self.defaultProxyFlushInterval;
	self.includeStatsdMetrics = false;
	self.includeInfluxdbMetrics = false;

	/* XXX Hardcoding default prefix here because it is not accessible otherwise. */
	self.prefixStats = config.prefixStats !== undefined ? config.prefixStats :
		'statsd';

	if (config.influxdb) {
		self.host = config.influxdb.host || self.defaultHost;
		self.port = config.influxdb.port || self.defaultPort;
		self.version = config.influxdb.version || self.defaultVersion;
		self.user = config.influxdb.username;
		self.pass = config.influxdb.password;
		self.database = config.influxdb.database;
		self.includeStatsdMetrics = config.influxdb.includeStatsdMetrics;
		self.includeInfluxdbMetrics = config.influxdb.includeInfluxdbMetrics;

		if (config.influxdb.ssl) {
			self.protocol = https;
		}

		if (config.influxdb.flush) {
			self.flushEnable = config.influxdb.flush.enable;
		}

		if (config.influxdb.proxy) {
			self.proxyEnable = config.influxdb.proxy.enable || self.defaultProxyEnable;
			self.proxySuffix = config.influxdb.proxy.suffix || self.defaultProxySuffix;
			self.proxyFlushInterval = config.influxdb.proxy.flushInterval || self.defaultProxyFlushInterval;
		}
	}

	if (self.version >= 0.9) {
		self.assembleEvent = self.assembleEvent_v09;
		self.httpPOST = self.httpPOST_v09;
	} else {
		self.assembleEvent = self.assembleEvent_v08;
		self.httpPOST = self.httpPOST_v08;
	}

	if (self.proxyEnable) {
		self.log('Starting the buffer flush interval. (every ' + self.proxyFlushInterval +
			'ms)');
		setInterval(function() {
			self.flushQueue();
		}, self.proxyFlushInterval);

		events.on('packet', function(packet, rinfo) {
			try {
				self.processPacket(packet, rinfo);
			} catch (e) {
				self.log(e);
			}
		});
	}

	if (self.flushEnable) {
		events.on('flush', function(timestamp, metrics) {
			try {
				self.processFlush(timestamp, metrics);
			} catch (e) {
				self.log(e);
			}
		});
	}

	events.on('status', function(writeCb) {
		for (var stat in self.influxdbStats) {
			writeCb(null, 'influxdb', stat, self.influxdbStats[stat]);
		}
	});

	return true;
}

function millisecondsSince(start) {
	diff = process.hrtime(start);
	return diff[0] * 1000 + diff[1] / 1000000;
}

InfluxdbBackend.prototype.log = function(msg) {
	util.log('[influxdb] ' + msg);
}

InfluxdbBackend.prototype.logDebug = function(msg) {
	if (this.debug) {
		var string;

		if (msg instanceof Function) {
			string = msg();
		} else {
			string = msg;
		}

		util.log('[influxdb] (DEBUG) ' + string);
	}
}

/**
 * Flush strategy handler
 *
 * @param {Number} timestamp
 * @param {Object} stats metric
 */
InfluxdbBackend.prototype.processFlush = function(timestamp, metrics) {
	var self = this,
		counters = metrics.counters,
		gauges = metrics.gauges,
		timerData = metrics.timer_data,
		statsdMetrics = metrics.statsd_metrics,
		points = [],
		sets = function(vals) {
			var ret = {};
			for (var val in vals) {
				ret[val] = vals[val].values();
			}
			return ret;
		}(metrics.sets),
		startTime = process.hrtime(),
		key, timerKey,
		statsPrefixRegexp = new RegExp('^' + self.prefixStats + '\\.');

	/* Convert timestamp from seconds to milliseconds. */
	timestamp = (timestamp * 1000);

	for (key in counters) {
		/* Do not include statsd counters. */
		if (!self.includeStatsdMetrics && key.match(statsPrefixRegexp)) {
			continue;
		}

		var value = counters[key],
			k = key + '.counter';

		if (value) {
			points.push(self.assembleEvent(k, [{
				value: value,
				time: timestamp
			}]));
		}
	}

	for (set in sets) {
		sets[set].map(function(v) {
			points.push(self.assembleEvent(set, [{
				value: v,
				time: timestamp
			}]));
		})
		points.push(self.assembleEvent(set + "_count", [{
			value: sets[set].length,
			time: timestamp
		}]));
	}

	for (key in gauges) {
		/* Do not include statsd gauges. */
		if (!self.includeStatsdMetrics && key.match(statsPrefixRegexp)) {
			continue;
		}

		var value = gauges[key],
			k = key + '.gauge';

		if (!isNaN(parseFloat(value)) && isFinite(value)) {
			points.push(self.assembleEvent(k, [{
				value: value,
				time: timestamp
			}]));
		}
	}

	for (key in timerData) {
		var timerMetrics = timerData[key];

		// Try to add histogram data, if it is there:
		if (timerMetrics.histogram) {
			var histoMetrics = timerMetrics.histogram,
				histoKey;

			for (histoKey in histoMetrics) {
				var value = histoMetrics[histoKey],
					k = key + '.timer.histogram.' + histoKey;

				points.push(self.assembleEvent(k, [{
					value: value,
					time: timestamp
				}]));
			}

			// Delete here so it isn't iterated over later:
			delete timerMetrics.histogram;
		}

		// Iterate over normal metrics:
		for (timerKey in timerMetrics) {
			var value = timerMetrics[timerKey],
				k = key + '.timer' + '.' + timerKey;

			points.push(self.assembleEvent(k, [{
				value: value,
				time: timestamp
			}]));
		}
	}

	if (self.includeStatsdMetrics) {
		// Include backend metrics for the previous flush
		if (self.includeInfluxdbMetrics) {
			statsdMetrics['influxdbStats.flush_time'] = self.influxdbStats.flushTime;
			statsdMetrics['influxdbStats.http_response_time'] = self.influxdbStats.httpResponseTime;
			statsdMetrics['influxdbStats.payload_size'] = self.influxdbStats.payloadSize;
			statsdMetrics['influxdbStats.num_stats'] = self.influxdbStats.numStats;
		}

		for (key in statsdMetrics) {
			var value = statsdMetrics[key],
				k = self.prefixStats + '.' + key;

			if (!isNaN(parseFloat(value)) && isFinite(value)) {
				points.push(self.assembleEvent(k, [{
					value: value,
					time: timestamp
				}]));
			}
		}
	}

	self.httpPOST(points);
	self.influxdbStats.flushTime = millisecondsSince(startTime);
}

InfluxdbBackend.prototype.processPacket = function(packet, rinfo) {
	var self = this,
		ts = (new Date()).valueOf();

	/* Stolen from statsd's stats.js. */
	var packet_data = packet.toString(),
		metrics;

	if (packet_data.indexOf("\n") > -1) {
		metrics = packet_data.split("\n");
	} else {
		metrics = [packet_data];
	}

	for (var midx in metrics) {
		if (metrics[midx].length === 0) {
			continue;
		}
		var bits = metrics[midx].toString().split(':');
		var key = bits.shift()
			.replace(/\s+/g, '_')
			.replace(/\//g, '-')
			.replace(/[^a-zA-Z_\-0-9\.]/g, '');

		if (bits.length === 0) {
			bits.push("1");
		}

		for (var i = 0; i < bits.length; i++) {
			var fields = bits[i].split("|");

			if (fields[1] === undefined) {
				self.log('Bad line: ' + fields + ' in msg "' + metrics[midx] + '"');
				continue;
			}

			var metric_type = fields[1].trim();

			/* Timer */
			if (metric_type === "ms") {
				self.enqueue('timer', ts, key, Number(fields[0] || 0));
				/* Gauge */
			} else if (metric_type === "g") {
				if (fields[0].match(/^[-+]/)) {
					self.logDebug('Sending gauges with +/- is not supported yet.');
				} else {
					self.enqueue('gauge', ts, key, Number(fields[0] || 0));
				}
				/* Set */
			} else if (metric_type === "s") {
				self.logDebug('Sets not supported yet.');
				/* Counter */
			} else {
				/* XXX Handle sampling. */
				self.enqueue('counter', ts, key, Number(fields[0] || 1));
			}
		}
	}
}

InfluxdbBackend.prototype.enqueue = function(type, ts, key, value) {
	var self = this;

	key = key + '.' + type + '.' + self.proxySuffix;

	if (!self.registry[key]) {
		self.registry[key] = [];
	}

	self.registry[key].push({
		value: value,
		time: ts
	});
}

InfluxdbBackend.prototype.flushQueue = function() {
	var self = this,
		registry = self.clearRegistry(),
		points = [];

	for (var key in registry) {
		var payload = self.assembleEvent(key, registry[key]);

		self.logDebug(function() {
			return 'Flush ' + registry[key].length + ' values for ' + key;
		});

		points.push(payload);
	}

	self.httpPOST(points);

	self.logDebug('Queue flushed');
}


InfluxdbBackend.prototype.clearRegistry = function() {
	var self = this,
		registry = self.registry;

	self.registry = {};

	return registry;
}

InfluxdbBackend.prototype.assembleEvent_v08 = function(name, events) {
	var self = this;

	var payload = {
		name: name,
		columns: Object.keys(events[0]),
		points: []
	};

	for (var idx in events) {
		var event = events[idx],
			points = [];

		for (var cidx in payload.columns) {
			var column = payload.columns[cidx];

			points.push(event[column]);
		}

		payload.points.push(points);
	}

	return payload;
}

InfluxdbBackend.prototype.assembleEvent_v09 = function(name, events) {
	var self = this;

	var payload = {
		name: name,
		timestamp: events[0]['time'],
		fields: {
			value: events[0]['value']
		}
	}

	return payload;
}

InfluxdbBackend.prototype.httpPOST_v08 = function(points) {
	/* Do not send if there are no points. */
	if (!points.length) {
		return;
	}

	var self = this,
		query = {
			u: self.user,
			p: self.pass,
			time_precision: 'ms'
		},
		protocolName = self.protocol == http ? 'HTTP' : 'HTTPS',
		startTime;

	self.logDebug(function() {
		return 'Sending ' + points.length + ' different points via ' +
			protocolName;
	});

	self.influxdbStats.numStats = points.length;

	var options = {
		hostname: self.host,
		port: self.port,
		path: '/db/' + self.database + '/series?' + querystring.stringify(query),
		method: 'POST',
		agent: false // Is it okay to use "undefined" here? (keep-alive)
	};

	var req = self.protocol.request(options);

	req.on('socket', function(res) {
		startTime = process.hrtime();
	});

	req.on('response', function(res) {
		var status = res.statusCode;

		self.influxdbStats.httpResponseTime = millisecondsSince(startTime);

		if (status !== 200) {
			self.log(protocolName + ' Error: ' + status);
		}
	});

	req.on('error', function(e, i) {
		self.log(e);
	});

	var payload = JSON.stringify(points)
	self.influxdbStats.payloadSize = Buffer.byteLength(payload);

	self.logDebug(function() {
		var size = (self.influxdbStats.payloadSize / 1024).toFixed(2);
		return 'Payload size ' + size + ' KB';
	});

	req.write(payload);
	req.end();
}

function csvKeyVal(obj) {
	var csv = [];
	for (var key in obj) {
		if (obj.hasOwnProperty(key)) {
			csv.push(key + "=" + obj[key]);
		}
	}

	return csv.join(",");
}

InfluxdbBackend.prototype.makeLineBasedPayload_v09 = function(points) {
	var fmtPayload = [];
	for (var idx in points) {
		var point = points[idx];
		var measurementName = point.name;
		var timestamp = point.timestamp;
		var fields = point.fields;

		var _fmtPayload = measurementName + " " + csvKeyVal(fields) + " " +
			timestamp;
		fmtPayload.push(_fmtPayload);
	}
	return fmtPayload.join("\n");
}

InfluxdbBackend.prototype.httpPOST_v09 = function(points) {
	/* Do not send if there are no points. */
	if (!points.length) {
		return;
	}

	var self = this,
		query = {
			u: self.user,
			p: self.pass,
			db: self.database,
			precision: 'ms'
		},
		protocolName = self.protocol == http ? 'HTTP' : 'HTTPS',
		startTime;

	self.logDebug(JSON.stringify(points))

	self.logDebug(function() {
		return 'Sending ' + points.length + ' different points via ' +
			protocolName;
	});

	self.influxdbStats.numStats = points.length;

	var options = {
		hostname: self.host,
		port: self.port,
		path: '/write?' + querystring.stringify(query),
		method: 'POST',
		agent: false // Is it okay to use "undefined" here? (keep-alive)
	};

	self.logDebug("url:" + options.hostname +
		":" + options.port + options.path);

	var req = self.protocol.request(options);

	req.on('socket', function(res) {
		startTime = process.hrtime();
	});

	req.on('response', function(res) {
		var status = res.statusCode;

		self.influxdbStats.httpResponseTime = millisecondsSince(startTime);

		if (status >= 400) {
			self.log(protocolName + ' Error: ' + status);
		}
	});

	req.on('error', function(e, i) {
		self.log(e);
	});

	var payload = JSON.stringify({
		database: self.database,
		time_precision: 'ms',
		points: points
	});

	// convert payload to new 0.9 format
	var payload = self.makeLineBasedPayload_v09(points);

	self.influxdbStats.payloadSize = Buffer.byteLength(payload);

	self.logDebug(function() {
		var size = (self.influxdbStats.payloadSize / 1024).toFixed(2);
		return 'Payload size ' + size + ' KB';
	});

	self.logDebug("Payload:" + payload);
	req.write(payload);
	req.end();
}

InfluxdbBackend.prototype.configCheck = function() {
	var self = this,
		success = true;

	/* Make sure the database name is configured. */
	if (!self.database) {
		self.log('Missing config option: database');
		success = false;
	}

	return success;
}

exports.init = function(startupTime, config, events) {
	var influxdb = new InfluxdbBackend(startupTime, config, events);

	return influxdb.configCheck();
}
